### PR TITLE
Travis improve

### PR DESCRIPTION
**Pull request by Soullivaneuh**
In a nutshell:

* composer vendor directory cache
* dist-mode for composer update
* GitHub oauth. You need to set a token on `GITHUB_OAUTH_TOKEN` variable. Mark it as private to secure it.
* Lowest deps mode to check composer dependencies going right.
* Some Travis optimizations for composer. Some time composer could be long to resolve and consume huge memory

I have some another optimizations idea but I'll talk about it later on another PR. ;-)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -8,23 +8,31 @@ php:
...
-    - php: hhvm
+sudo: false
 
 cache:
   directories:
     - $HOME/.composer/cache
+    - vendor
+
+matrix:
+  fast_finish: true
```
Comment (ronanguilloux) [inline .travis.yml]: Pirate change. See https://github.com/ronanguilloux/IsoCodes/pull/26

Comment (Soullivaneuh) [inline .travis.yml]: Indeed my bad. If I'm right it should be added on a separate PR for discussion. Will do it later.

**Comment by Soullivaneuh** [comment]
Review fixed.

**Comment by Soullivaneuh** [comment]
Had to remove `bin-dir` option because bin are not re-installed if vendor cache is not updated.

**Comment by Soullivaneuh** [comment]
@ronanguilloux Can you please clean Travis cache and re-launch this PR?

**Comment by Soullivaneuh** [comment]
Yes, works great now. Ready to merge.

File: .travis.yml [hunk 2]
```diff
@@ -8,26 +8,33 @@ php:
...
 
 cache:
   directories:
     - $HOME/.composer/cache
+    - vendor
+
+matrix:
+  include:
+    - php: 5.4
+      env: COMPOSER_FLAGS="--prefer-lowest"
```
Comment (ronanguilloux) [inline .travis.yml]: Sorry, I don't get the point about fetching and testing *minimal* versions of requirements: could you please elaborate on that choice?

Comment (Soullivaneuh) [inline .travis.yml]: --prefer-lowest will get the lowest dependencies base on you composer configuration.

In this case, we can prevent using methods/tools from dependencies that older versions does not have.

Basically, it's will prevent from created issues like "Your library throw an undefined method error"  and discover it's because user has not enough higher dependencies, because composer will tell him before that.

Many project use this system. See Symfony for example: https://github.com/symfony/symfony/blob/dd7583d4851fa10f711b3f7e0727f80236b22f77/.travis.yml#L16-L17

File: composer.json [hunk 3]
```diff
@@ -36,9 +36,6 @@
     "autoload": {
         "psr-0": {"IsoCodes": "src/"}
     },
-    "config": {
-        "bin-dir": "bin"
```
Comment (ronanguilloux) [inline composer.json]: Makefile's prepared tasks use a few quality-related tools: phpcs, php-cs-fixer, phploc, phpmd, etc. Composer actually (symlink-)install them in that `/bin` dir, so far. If you change that actual vendors bin directory path, they cannot work anymore.

File: .travis.yml [hunk 4]
```diff
@@ -8,26 +8,33 @@ php:
...
+  allow_failures:
+    - php: 7.0
+    - php: hhvm
 
 before_script:
-  - composer self-update
-  - composer install --prefer-source --no-interaction
```
Comment (ronanguilloux) [inline .travis.yml]: - OK for changing `--prefer-dist`: Maybe `--prefer-dist` would be a better choice.
- `--no-intercation` makes sense, I wouldn't have removed it.

File: composer.json (hunk 3 above)
Comment (Soullivaneuh) [inline composer.json]: I updated the bin path on script too.

I had to do this. See: https://github.com/ronanguilloux/IsoCodes/pull/57#issuecomment-114264368

BTW, I have another suggestion to do about this on another PR.

**Comment by Soullivaneuh** [comment]
@ronanguilloux Please remember to set `GITHUB_OAUTH_TOKEN` before merge this request.

File: .travis.yml [hunk 5]
```diff
@@ -8,26 +8,33 @@ php:
...
+  - mkdir -p ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d && echo "memory_limit=-1" >> ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/travis.ini
+  - composer selfupdate
+  - composer config -q github-oauth.github.com $GITHUB_OAUTH_TOKEN
+  - travis_wait composer update --prefer-dist --no-interaction $COMPOSER_FLAGS
 
 script:
-  - mkdir -p build/logs
-  - mkdir -p build/cov
-  - php bin/phpunit -d error_reporting=16384
```
Comment (ronanguilloux) [inline .travis.yml]: 16384  = User-generated warning message. This is like an E_DEPRECATED, except it is generated in PHP code by using the PHP function trigger_error(). See (php doc about errorfunc constants](http://php.net/manual/en/errorfunc.constants.php). 
This error-reporting conf was introduced to fix #45.

Comment (Soullivaneuh) [inline .travis.yml]: I removed because I plan to propose a better option with `symfony/phpunit-bridge`.

See https://github.com/ronanguilloux/IsoCodes/issues/58.

Do you want me to integrate on this PR directly?

File: .travis.yml [hunk 6]
```diff
@@ -8,26 +8,33 @@ php:
...
+      env: COMPOSER_FLAGS="--prefer-lowest"
+  allow_failures:
+    - php: 7.0
+    - php: hhvm
 
 before_script:
-  - composer self-update
-  - composer install --prefer-source --no-interaction
+  - mkdir -p ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d && echo "memory_limit=-1" >> ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/travis.ini
+  - composer selfupdate
```
Comment (ronanguilloux) [inline .travis.yml]: Re-reading Travis-CI documentation, you may found that

> (...) we update composer every time we update the PHP build environment, which is every 30-60 days. Because composer has a time-based update warning, you may see messages (related to `30 days old` composer build), which may be safely ignored.

We may omit that step so far.

Comment (Soullivaneuh) [inline .travis.yml]: I ensure you by experience that Composer is not updated every 30-60 days. ;-)

I have to use and see this approach on many projects to get latest version of composer directly.

This is not a high consuming job, but I can removed it if you absolutely want.

What do you choose?

File: .travis.yml [hunk 7]
```diff
@@ -8,26 +8,33 @@ php:
...
+  - composer config -q github-oauth.github.com $GITHUB_OAUTH_TOKEN
+  - travis_wait composer update --prefer-dist --no-interaction $COMPOSER_FLAGS
```
Comment (ronanguilloux) [inline .travis.yml]: Isn't this duplicating the `composer install --prefer-source --no-interaction` command in line 22?

File: .travis.yml [hunk 8]
```diff
@@ -8,26 +8,33 @@ php:
...
 
 script:
-  - mkdir -p build/logs
-  - mkdir -p build/cov
-  - php bin/phpunit -d error_reporting=16384
+  - mkdir -p build/logs build/cov
+  - php vendor/bin/phpunit
```
Comment (ronanguilloux) [inline .travis.yml]: See note about lines 27 above and 40 below.

File: .travis.yml (hunk 7 above)
Comment (Soullivaneuh) [inline .travis.yml]: The 22 line is a deleted line.

File: .travis.yml (hunk 8 above)
Comment (Soullivaneuh) [inline .travis.yml]: I answered on this note.

File: .travis.yml [hunk 9]
```diff
@@ -8,26 +8,33 @@ php:
...
 
 after_script:
-  - php bin/coveralls -v
+  - php vendor/bin/coveralls -v
```
Comment (ronanguilloux) [inline .travis.yml]: See note about line 40 below.

File: .travis.yml (hunk 5 above)
Comment (ronanguilloux) [inline .travis.yml]: Please keep PR granularity principle: if this change is related to another PR, remove it from this one and attach it to this other PR.

File: .travis.yml (hunk 9 above)
Comment (Soullivaneuh) [inline .travis.yml]: https://github.com/ronanguilloux/IsoCodes/pull/57#discussion_r33034296

File: .travis.yml (hunk 5 above)
Comment (Soullivaneuh) [inline .travis.yml]: Fixed.

**Comment by Soullivaneuh** [comment]
@ronanguilloux I answered on all diff comments.

Waiting for an answer on this one: https://github.com/ronanguilloux/IsoCodes/pull/57#discussion_r33034890

Regards.

File: .travis.yml (hunk 6 above)
Comment (ronanguilloux) [inline .travis.yml]: Let's rely on Travis-CI documentation: please skip composer self-update.

File: .travis.yml (hunk 7 above)
Comment (ronanguilloux) [inline .travis.yml]: OK. I keep composer.lock updated => please use `composer install` instead

Comment (Soullivaneuh) [inline .travis.yml]: No. Using `composer install` will avoid all the interest of testing a library.

Installed as a library via composer, users won't be synced to your lock file.

Libraries must be tested every time with both lowest and latest dependencies to cover all cases.

composer.lock should not be here. It's only useful for project, not library. But I think you had scrutinizer issue without it...

BTW, `composer install` will fail with `--prefer-lowest` option.

All library projects I see (except this one) use `composer update` for Travis test.

Regards.

File: .travis.yml (hunk 6 above)
Comment (Soullivaneuh) [inline .travis.yml]: Ok I'll do it. I just want to advice that this doc part is an indication, not an obligation. ;-)

File: .travis.yml (hunk 2 above)
Comment (ronanguilloux) [inline .travis.yml]: Got it. [Further reading found here](http://evertpot.com/testing-composer-prefer-lowest).
OK for that.

File: composer.json (hunk 3 above)
Comment (ronanguilloux) [inline composer.json]: Makefile changes wasn't commited here, `make` commands won't work anymore.

Comment (Soullivaneuh) [inline composer.json]: Didn't see the `Makefile`, will take a look.

BTW, why `make` is not use on Travis? :-)

File: .travis.yml [hunk 10]
```diff
@@ -8,26 +8,32 @@ php:
...
-matrix:
-  allow_failures:
-    - php: 7.0
-    - php: hhvm
+sudo: false
 
 cache:
   directories:
     - $HOME/.composer/cache
+    - vendor
```
Comment (ronanguilloux) [inline .travis.yml]: I don't get the point of caching vendors and running a `composer update`. The cache’s purpose is to make installing language-specific dependencies easy and fast, but our concern is to test the codebase against many changing vendors dependencies.

This also has an unresolved side effect over the `Makefile`. 

In addition, & I don't want to be asked to manually clear the Travis cache every time a PR encounter a BC break issue and the PR author wants to digg into a reliable travis job trace.

For now, I'd suggest not to cache the vendors, unless we found a reason to do so.
I'm open to any clear demonstration about an opposite POV.

Comment (Soullivaneuh) [inline .travis.yml]: > The cache’s purpose is to make installing language-specific dependencies easy and fast

Right.

> but our concern is to test the codebase against many changing vendors dependencies.

Wrong, or I don't get it. Vendor caching is here just to avoid re-installation of **already up to date** dependencies according to `composer.json` file.

Caching vendor or not, the behavior will be the same: All dependencies will be updated if necessary.

The advantage is that only needed package are changed, that make going Travis faster.

> This also has an unresolved side effects over the `Makefile` tasks.

Which one? And BTW, Makefile is not used on Travis AFAIK.

> In addition, as a maintainer, I don't want to be asked to manually clear the Travis cache every time a PR encounter a BC break issue and the PR author wants to digg into a reliable travis job trace.

Cache clearing would not be needed. I ask for it here because of `dir` bin issue. It's a special case.


I hope I'm enough clear on my explanation. Don't hesitate to ask me if not.

Comment (ronanguilloux) [inline .travis.yml]: > All dependencies will be updated if necessary. 

Didn't find the information on how this actually works in Travis doc. 

> Which one? And BTW, Makefile is not used on Travis AFAIK.

Travis isn't concerned here, but contributors/hackers are. Please read the `README.md` to understand what the `Makefile` stands for. Your changes on `bin` dir actually makes the `Makefile` bugging.

Comment (Soullivaneuh) [inline .travis.yml]: > Didn't find the information on how this actually works in Travis doc.

This is not about Travis but about composer itself.

Just run a `composer update` on your project, change a deps version or include a new one on your `composer.json` and run `composer update` again. You will see that only the concerned package will be changed.
This is the same behavior on Travis.

> Your changes on bin dir actually makes the Makefile bugging.

Well, this is related to bin dir configuration so, not cache.

As I said, I'll take a look for that. ;-)

Comment (Soullivaneuh) [inline .travis.yml]: I'll finally remove vendor cache because of this issue: https://github.com/composer/composer/issues/4147#issuecomment-115258930

**Comment by Soullivaneuh** [comment]
@ronanguilloux Reviews fixed and tests passed. Ready to merge. :+1:

**Comment by Soullivaneuh** [comment]
ping @ronanguilloux
